### PR TITLE
Upgrade InfluxDB storage to InfluxDB 0.9

### LINKED ISSUE #743: InfluxDB v0.9 support
Are there any plans to support the new InfluxDB version? 0.9 isn't officially out yet, but it seems like it should be done soon, and will be the first stable production release. Unfortunately the API that cadvisor uses to connect  has changed drastically since the last supported version, 0.8.

### PR DESCRIPTION
**Pull request by mnuessler**
- Fix #743
- Update InfluxDB client dependency to version 0.9.1-rc1.
- Rewrite InfluxDB storage for new InfluxDB API data structures.
- Store each measurement separately instead of storing all measurements
  in a single big "table" with many columns/fields.
- Use tags add metadata to points, such as the container name. Tags are
  a new feature in InfluxDB 0.9.

### CODE REVIEW

**Comment by mnuessler** [comment]
Not sure why the CI build failed. I can't reproduce the problem locally. Looks like dependency is missing?

**Comment by vmarmol** [comment]
Hmmm the diff is so bit GitHub is choking on it :( can we split the PR into two? Let's try to get the Godeps in first and then the actual changes.

**Comment by mnuessler** [comment]
Ok, I split the commit.

**Comment by benfb** [comment]
Any updates on this?

**Comment by mnuessler** [comment]
@vmarmol Any feedback on this PR so far? :)

**Comment by mikedanese** [comment]
@mnuessler The godep changes broke the build.

```
godep go test -v -race -test.short github.com/google/cadvisor/...
utils/cloudinfo/gce.go:20:2: cannot find package "github.com/GoogleCloudPlatform/gcloud-golang/compute/metadata" in any of:
	/home/travis/.gimme/versions/go1.4.linux.amd64/src/github.com/GoogleCloudPlatform/gcloud-golang/compute/metadata (from $GOROOT)
	/home/travis/gopath/src/github.com/google/cadvisor/Godeps/_workspace/src/github.com/GoogleCloudPlatform/gcloud-golang/compute/metadata (from $GOPATH)
	/home/travis/gopath/src/github.com/GoogleCloudPlatform/gcloud-golang/compute/metadata
storage/bigquery/client/example/example.go:22:2: cannot find package "github.com/SeanDolphin/bqschema" in any of:
	/home/travis/.gimme/versions/go1.4.linux.amd64/src/github.com/SeanDolphin/bqschema (from $GOROOT)
	/home/travis/gopath/src/github.com/google/cadvisor/Godeps/_workspace/src/github.com/SeanDolphin/bqschema (from $GOPATH)
	/home/travis/gopath/src/github.com/SeanDolphin/bqschema
utils/fs/mockfs/mockfs.go:21:2: cannot find package "code.google.com/p/gomock/gomock" in any of:
	/home/travis/.gimme/versions/go1.4.linux.amd64/src/code.google.com/p/gomock/gomock (from $GOROOT)
	/home/travis/gopath/src/github.com/google/cadvisor/Godeps/_workspace/src/code.google.com/p/gomock/gomock (from $GOPATH)
	/home/travis/gopath/src/code.google.com/p/gomock/gomock
```

**Comment by mnuessler** [comment]
Can someone with more experience in godep please give me a hint on how to fix that? I'm kind of stuck here...

**Comment by mnuessler** [comment]
I finally managed to fix the broken dependencies. Please have a look at this PR, thank you.

File: Godeps/Godeps.json [hunk 1]
```diff
@@ -1,9 +1,6 @@
 {
 	"ImportPath": "github.com/google/cadvisor",
 	"GoVersion": "go1.4.2",
-	"Packages": [
-		"./..."
-	],
```
Comment (mikedanese) [inline Godeps/Godeps.json]: This should not be removed

File: Godeps/Godeps.json [hunk 2]
```diff
@@ -117,28 +123,65 @@
...
+			"Comment": "v0.9.1",
+			"Rev": "8b3219e74fcc3843a6f4901bdf00e905642b6bd6"
 		},
 		{
 			"ImportPath": "github.com/kr/pretty",
 			"Comment": "go.weekly.2011-12-22-20-g088c856",
 			"Rev": "088c856450c08c03eb32f7a6c221e6eefaa10e6f"
+			"ImportPath": "github.com/influxdb/influxdb/influxql",
+			"Comment": "v0.9.1",
+			"Rev": "8b3219e74fcc3843a6f4901bdf00e905642b6bd6"
```
Comment (mikedanese) [inline Godeps/Godeps.json]: These imports are not valid json. Different imports need to be in separate objects.

**Comment by mikedanese** [comment]
How are you updating godeps? Can you explain your process?

**Comment by mnuessler** [comment]
@mikedanese  I have fixed the Godeps.json according to your comments.

I tried several ways of updating godeps, but none seem to work as I expected. Maybe you could point me to the correct way.

Actually I just need to upgrade the package `github.com/influxdb/influxdb/client` to the latest release (v0.9.1 or at least anything > v0.9.0). Starting off with a clean GOPATH, I tried the following steps:

*  Fetch cadvisor via `go get github.com/google/cadvisor`.
* Restore dependencies in GOPATH with `godep restore`. (Not sure if this is actually necessary.)
* Use git to pull my cadvisor branch and switch to it.
* Update the influxdb package in the GOPATH: `go get -u github.com/influxdb/influxdb`
* Now when trying to update the godeps with `godeps save` or `godeps save ./...` I get an error like such as:
```
godep: github.com/influxdb/influxdb/influxql: revision is 8b3219e74fcc3843a6f4901bdf00e905642b6bd6, want 3284662b350688b651359f9124928856071bd3f5
```
* When I do `godep save github.com/influxdb/influxdb/client` or ``godep save github.com/influxdb/influxdb` then godeps are updated, but a lot of dependencies also get deleted which caused the CI build to fail.

The only way I could get it to work with the CI was committing the added godeps but ignoring the deletions. It worked locally because the missing dependencies were still in the GOPATH. Any idea what I am doing wrong here?

**Comment by SlashmanX** [comment]
Any progress on getting this merged?

**Comment by mnuessler** [comment]
Any comments on this? It's been a month already since I opened this PR...

**Comment by rjnagal** [comment]
IIRC, after merging this, we end up dropping support for older influxdb formats. Is that right?
We have quite a few users at the older version. @vishh do you think we need to support both versions for a while, or is it okay to drop the old one?

**Comment by rjnagal** [comment]
Discussed offline. It's fine to drop the older version support.

I took another shot at reviewing this. I can't make sure if we need all these dependencies (soo many!).
The storage driver changes look okay though.

I'll setup influxdb 0.9 and give it another try. Thanks for your patience!

**Comment by ClanStyles** [comment]
Any update? I'd love to push this into prod, CAdvisor is one of the last reasons we haven't pushed Influx 0.9. We've begun looking into telegraf.

**Comment by rjnagal** [comment]
@vishh Do you want to try this out? I am fire-fighting this week :sob:

**Comment by vishh** [comment]
Sure. Will review soon.

**Comment by hectorj2f** [comment]
@vishh Do we have any news on this ?

**Comment by selvik** [comment]
Looking forward to this feature getting in. Thanks!

**Comment by vishh** [comment]
@mnuessler: I'm not able to review this patch using github. Can you remove the godep changes temporarily? Once the review is complete, you can add it back.

**Comment by mnuessler** [comment]
@vishh I already split the change into two commits for that reason. There is one commit containing only the godep changes and another one containing only the code changes. Please have a look, maybe that is sufficient?

Unfortunately the InfluxDB client pulls in a plethora of additional dependencies...

**Comment by vishh** [comment]
Can you squash all non godeps changes into a single commit? It will help
having to juggle between commits.

**Comment by mnuessler** [comment]
@vishh Ok, squashed commits as suggested.

**Comment by vishh** [comment]
Just a few nits. Overall LGTM. Thanks @mnuessler for working on this! And apologies for the slow review! Do you care to post an upgrade guide? Upgrading to 0.9 is non-trivial AFAIK.

**Comment by mikedanese** [comment]
The huge number of dependencies for the client is actually fixed in HEAD of influx.
https://github.com/influxdb/influxdb/pull/4134
https://github.com/influxdb/influxdb/issues/3447

**Comment by mnuessler** [comment]
@mikedanese Looks promising! I will do a test with the InfluxDB client from HEAD.

**Comment by mnuessler** [comment]
@vishh Thanks for the review! Will reply to your comments tomorrow.

**Comment by mangalaman93** [comment]
Why doesn't influxdb show any network interfaces? Is that the expected behavior? I was trying to use these changes in my cadvisor-influxdb-grafana setup.

**Comment by vishh** [comment]
There was an issue with network statistics which has been fixed on HEAD.
The next release of cadvisor should surface network stats without any
issues.

**Comment by vishh** [comment]
@mnuessler: Any updates on this PR?

**Comment by mnuessler** [comment]
@vishh Using the latest InfluxDB client from master we really don't need all those extra dependencies anymore, that's nice. I still need to do a test for some changes according to your comments.

**Comment by onlyjob** [comment]
Do not ever use anything "latest", "from master" etc. Please stick to formal release. See [list of InfluxDB releases](https://github.com/influxdb/influxdb/releases). The latest _released_ version of InfluxDB is [v0.9.4.2](https://github.com/influxdb/influxdb/tree/v0.9.4.2).

**Comment by mnuessler** [comment]
@onlyjob Unfortunately the client of the latest release still requires that huge number of extra dependencies. If we cannot use the latest master and want to avoid pulling in all those dependencies, then we have to wait for the 0.9.5 release.

**Comment by onlyjob** [comment]
I'm not concerned about extra dependencies. But if v0.9.4.2 is not good enough then by all means wait till 0.9.5 please. Release number are there precisely to define dependencies as well as to use software that is less likely to break comparing to unreleased development version from repository.
Depending on unreleased software (and even worse, bundling it) is a nightmare for downstream package maintainers. Please don't do that.

**Comment by mnuessler** [comment]
@mikedanese @vishh What do you think? Should we use 0.9.4.2 then and pull in all those dependencies?

**Comment by jimmidyson** [comment]
Personally I'd rather wait until dependencies are reduced with 0.9.5. cadvisor is embedded inside a number of other systems, including kubernetes, & the amount of dependencies could very well be a limiting factor in upgrading there.

Does anyone know of a timescale for influxdb 0.9.5?

**Comment by onlyjob** [comment]
I hope they do not release on schedule but given frequency of previous releases and the fact that the last release was about 20 days ago the new release might be just within weeks from now. Hopefully that's not too long and I also hope there is no rush to commit pre-release here...

**Comment by vishh** [comment]
One option until influxDB v0.9.5 is released is to maintain a separate
branch for InfluxDB v0.9 support. Will that be helpful?

**Comment by onlyjob** [comment]
From downstream maintainer prospective a separate branch based on pre-release software would not be harmful as long as it is not used to produce a tagged release...

**Comment by jimmidyson** [comment]
@onlyjob Any chance of updating this with latest influxdb client changes?

**Comment by SlashmanX** [comment]
InfluxDB v0.9.5 has been released, which should help speed this up, correct?

**Comment by onlyjob** [comment]
@jimmidyson I'm not sure why you are asking me... Pull request is not mine and I'm not qualified to update it nor have any time for it...

**Comment by jimmidyson** [comment]
@onlyjob Sorry - copy/pasted wrong person. Ignore me!

@SlashmanX It should.

@mnuessler Now that influxdb 0.9.5 is released could you update godeps please?

**Comment by mnuessler** [comment]
Thanks for the info. I will update the PR for InfluxDB 0.9.5. So we can rid of the extra dependencies now and still refer to a regular release.

**Comment by jimmidyson** [comment]
Thanks!

**Comment by vishh** [comment]
@mnuessler: Any updates?

**Comment by christianhuening** [comment]
Is there anything one could do to help getting this done? I've not looked into the sources or anything, but if it's merely a case of updating stuff and checking it, I might be able to do so.

**Comment by carmark** [comment]
@mnuessler 

I would like to help you do something if you are busy on other things.  It seems this patch is aspired by many guys in https://github.com/google/cadvisor/issues/743.

**Comment by mnuessler** [comment]
Sorry for delay. I managed to upgrade the InfluxDB client dependency to v0.9.5.1, but I had to rewrite commits in this branch. I also haven't tested it yet and I won't have time to do so before Monday.

Thanks @christianhuening and @carmark for your offer to help. I'm afraid I need to have a look first what still has to be done as I haven't worked on this for a while. But feel free to have a look at the code or do some testing if you like.

**Comment by mnuessler** [comment]
Rebased on master and resolved merge conflicts due to recent storage driver refactoring.

**Comment by jimmidyson** [comment]
Thanks for the update.

Travis showing failing gofmt check:

```
The following files are not properly formatted:
storage/influxdb/influxdb.go storage/influxdb/influxdb.go
```

**Comment by marcellodesales** [comment]
@mnuessler @christianhuening @carmark Is there a way to revisit this PR and get it fixed? Is there a way to help?

**Comment by jimmidyson** [comment]
Let me pull this down locally & submit a separate PR tomorrow with these changes & any necessary fixes.

@marcellodesales @carmark Feel free to do code review if you have time.

**Comment by christianhuening** [comment]
sorry was very busy over the holidays. I understand that the majority of the work has been done?

**Comment by marcellodesales** [comment]
@jimmidyson Thanks a lot for getting this! I hope I can do any code review after a  PR that passes tests!

**Comment by carmark** [comment]
@jimmidyson Thanks, it is my pleasure to do the code review.

**Comment by normanjoyner** [comment]
Submitted [PR](https://github.com/egymgmbh/cadvisor/pull/2) to fix missing InfluxDB models dependency. Tests failing on `TestGetProcessNamePid` which is fixed in dee3f07b0afc36bd41014679989e9610a4c1e8a7.

**Comment by jimmidyson** [comment]
This has now been superseded by #1040, including the commits from this PR & cleaning up a few things (including missing godeps). Closing this but please review/comment on #1040 so we can get that merged ASAP.

**Pull request closed without merging**